File: src/ai/node_watcher/src/NodeStatusHandler.cpp
#include "node_watcher/NodeStatusHandler.hpp"

NodeStatusHandler::NodeStatusHandler()
    : nh(), waitRequestCode(-1) {
    // Init node_watcher service client
    waiterClient = nh.serviceClient<AwaitNodesRequest>(Topics::NODES_AWAITER_INIT_SERVICE);
    watcherClient = nh.serviceClient<NodeReadiness>(Topics::NODE_WATCHER_SERVICE);
    
    // Send signal to start waiting
    startWaitPub = nh.advertise<std_msgs::Int32>(Topics::NODES_AWAITER_START_TOPIC, 10);
    
    // Listen for answers
    waitAnswerSub = nh.subscribe(Topics::NODES_AWAITER_RESULT_TOPIC, 10, &NodeStatusHandler::onAwaitResponse, this);

    // Wait for service to send init signal
    ros::service::waitForService(Topics::NODE_WATCHER_SERVICE);
    
}

string NodeStatusHandler::makeNodePath(string nodename, string package) {
    return "/" + package + "/" + nodename;
}

/**
 * Set given node status 
 */
void NodeStatusHandler::setNodeStatus(string nodename, string package, int state_code, int errorCode /*= 0*/) {
    this->setNodeStatus(this->makeNodePath(nodename, package), state_code, errorCode);
}

/**
 * Set given node status 
 */
void NodeStatusHandler::setNodeStatus(string nodepath, int state_code, int errorCode /*= 0*/) {
    // update remote version
    NodeReadiness msg;
    msg.request.status.state_code = state_code;
    msg.request.status.error_code = errorCode;
    msg.request.node_name = nodepath;

    if (watcherClient.call(msg)) {
        return;
    }

    ROS_ERROR_STREAM(
        "Unable to call watcher as setter for " << nodepath
    );
}

/**
 * Get an other node's status, with it's package name provided
 */
NodeStatus NodeStatusHandler::getNodeStatus(string nodename, string package) {
    return getNodeStatus(this->makeNodePath(nodename, package));
}

/**
 * Get an other node's status
 */
NodeStatus NodeStatusHandler::getNodeStatus(string nodepath) {
    NodeReadiness msg;
    msg.request.status.state_code = NodeStatus::NODE_ASKING; // ask for status
    msg.request.node_name = nodepath;

    if (watcherClient.call(msg)) {
        return msg.response.status;
    }

    ROS_ERROR_STREAM(
        "Unable to call watcher as getter for " << nodepath
    );
}

/**
 * Wait for required nodes to be alive
 */
void NodeStatusHandler::waitForNodes(int timeout) {
    if (requirements.size() == 0) {
        this->waitCallback(true);
    }

    // Prepare request
    AwaitNodesRequest request;
    request.request.nodes = requirements;
    request.request.timeout = timeout;
    
    // Call service to retrieve code
    if (waiterClient.call(request)) {
        this->waitRequestCode = request.response.request_code;

        // Wait for publisher to be ready to address not sent issue
        // From [https://answers.ros.org/question/11167/how-do-i-publish-exactly-one-message/]
        ros::Rate poll_rate(100);
        while(startWaitPub.getNumSubscribers() == 0) {
            poll_rate.sleep();
        }

        std_msgs::Int32 msg;
        msg.data = this->waitRequestCode;
        startWaitPub.publish(msg);
    } else {
        ROS_ERROR_STREAM("Unabled to call nodes waiting service");
        this->waitCallback(false);
    }
}

/**
 * Set the callback for waiting
 */
void NodeStatusHandler::setWaitCallback(WaitCallback cb) {
    this->waitCallback = cb;
}

void NodeStatusHandler::onAwaitResponse(const AwaitNodesResult& msg) {
    if (msg.request_code == this->waitRequestCode) {
        this->waitCallback(msg.success);
    }
}

/**
 * Add given node to required nodes
 */
void NodeStatusHandler::require(std::string nodename, std::string package, bool required /*= true*/) {
    this->require(this->makeNodePath(nodename, package), required);
}

/**
 * Add given node to required nodes
 */
void NodeStatusHandler::require(std::string nodepath, bool required /*= true*/) {
    NodeRequirement req;
    req.nodename = nodepath;
    req.optional = !required;

    this->requirements.push_back(req);
}

/**
 * Add nodes contained in given file to required nodes
 * 
 * The file should respect the following format (one node per line):
 * [nodename] required|optional
 */
void NodeStatusHandler::requireFromFile(std::string filename) {
    std::ifstream filestream;
    
    filestream.open(filename);

    if (filestream.fail()) {
        ROS_ERROR_STREAM("Unable to load dependency file : " + filename);
        this->waitCallback(false);
        return;
    }

    // Parse file
    string nodename;
    string status;
    int linenumber = 1;

    // First element of line parsed as nodename
    while(std::getline(filestream, nodename, ' ')) {
        // Then the node optional argument
        if (std::getline(filestream, status)) {
            // Check for error
            if (status != "optional" && status != "required") {
                ROS_WARN_STREAM("parsing requirements from " << filename << ": " << status
                    << " is not a valid optional argument (optional or required)");
            }

            // Save as required
            this->require(nodename, status == "required");
        } else {
            ROS_WARN_STREAM("parsing requirements from " << filename << ": line " << linenumber
                << " ignored due to lack of args");
        }

        linenumber += 1;
    }
}

bool NodeStatusHandler::isRequired(string nodepath) {
    for (const auto& nextReq : this->requirements) {
        if (nextReq.nodename == nodepath) {
            return true;
        }
    }

    return false;
}

File: src/ai/node_watcher/include/node_watcher/node_status_handler.py
import rospy
import rospkg

from ai_msgs.msg import AwaitNodesResult, \
	NodeStatus, NodeRequirement, Topics
from ai_msgs.srv import NodeReadiness, NodeReadinessRequest, AwaitNodesRequest

from std_msgs.msg import Int32

from typing import List, Callable

WaitCallback = Callable[[bool], None]

class NodeStatusHandler:
	'''Class handling any "abstract nodes" (as they are not defined as a single class)'''

	def __init__(self):
		self._watcher_client = rospy.ServiceProxy(Topics.NODE_WATCHER_SERVICE, NodeReadiness)
		self._waiter_client = rospy.ServiceProxy(Topics.NODES_AWAITER_INIT_SERVICE, AwaitNodesRequest)
		self._start_wait_pub = rospy.Publisher(Topics.NODES_AWAITER_START_TOPIC, Int32, queue_size=10)
		self._wait_answer_sub = rospy.Subscriber(Topics.NODES_AWAITER_RESULT_TOPIC, AwaitNodesResult, self._on_await_response)
	
		self._wait_callback: WaitCallback = None
		self._wait_request_code: int = -1

		self._requirements: List[NodeRequirement] = []

		# Wait for service to send init signal
		rospy.wait_for_service(Topics.NODE_WATCHER_SERVICE)

	# Build node path
	def _make_node_path(self, nodename: str, package: str) -> str:
		return "/{}/{}".format(package, nodename)

	# Status getter
	def get_node_status(self, nodename: str, package: str = None) -> NodeStatus:
		if package == None:
			nodename = self._make_node_path(nodename, package)

		msg = NodeReadinessRequest()
		msg.status.state_code = NodeStatus.NODE_ASKING # ask for status
		msg.node_name = nodename

		try:
			response = self._watcher_client(msg)
			return response.status

		except rospy.ServiceException:
			rospy.logerr("Unable to call watcher as getter for {}".format(nodename))
	
	# Status setter
	def set_node_status(self, nodename: str, package: str, state_code: int, error_code: int = 0):
		nodepath = self._make_node_path(nodename, package)

		# update remote version
		msg = NodeReadinessRequest()
		msg.status.state_code = state_code
		msg.status.error_code = error_code
		msg.node_name = nodepath

		try:
			self._watcher_client(msg)
		except rospy.ServiceException:
			rospy.logerr("Unable to call watcher as setter for {}".format(nodepath))
		

	# Node waiting
	def wait_for_nodes(self, timeout: int):
		if len(self._requirements) == 0:
			self._wait_callback(True)
		
		# Call service to retrieve code
		try:
			response = self._waiter_client(self._requirements, timeout)
			self._wait_request_code = response.request_code

			# Wait for publisher to be ready to address not sent issue
			# From [https:#answers.ros.org/question/11167/how-do-i-publish-exactly-one-message/]
			poll_rate = rospy.Rate(100)
			while self._start_wait_pub.get_num_connections() == 0:
				poll_rate.sleep()
			
			# Start waiting
			msg = Int32()
			msg.data = self._wait_request_code
			self._start_wait_pub.publish(msg)
		except rospy.ServiceException:
			rospy.logerr("Unabled to call nodes waiting service")
			self._wait_callback(False)

	# Callback setter
	def set_wait_callback(self, cb: WaitCallback):
		self._wait_callback = cb

	# Add node requirement
	def require_from_file(self, filename: str):
		try:
			file = open(filename, "r")

			for line in file:
				nodename, status = line.split(" ", 2)

				# Invalid status
				if status != "optional" and status != "required":
					rospy.logerr(
						"parsing requirements from {} : {} is not a valid optional argument (optional or required)"
							.format(filename, status)
					)
				else:
					# Save as required
					self.require(nodename, status == "required")
		except:
			rospy.logerr("unable to parse {}".format(filename))

		
	def require(self, nodename: str, package: str, required: bool = True):
		nodepath = self._make_node_path(nodename, package)

		req = NodeRequirement()
		req.nodename = nodepath
		req.optional = not required

		self._requirements.append(req)

	# Requirement getter
	def is_required(self, nodepath: str) -> bool:
		for next in self._requirements:
			if next.nodename == nodepath:
				return True

		return False

	def _on_await_response(self, msg: AwaitNodesResult):
		if msg.request_code == self._wait_request_code:
			self._wait_callback(msg.success)
